### PR TITLE
Fix sync failure on unreachable namespace avatar

### PR DESCRIPTION
**Pull request by gerrod3**
fixes: #1543

### CODE REVIEW

File: pulp_ansible/app/tasks/collections.py [hunk 1]
```diff
@@ -1074,6 +1074,25 @@ async def run(self):
...
+
+    async def download(self):
+        """Allow download to fail, but not stop the sync."""
+        artifact_copy = self.artifact
+        try:
+            return await super().download()
+        except ClientError as e:
+            # Reset DA so that future stages can properly handle it
+            self.artifact = artifact_copy
+            self.deferred_download = True
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collections.py:1090]: It reminds me a lot of https://github.com/pulp/pulp_deb/blob/main/pulp_deb/app/tasks/synchronizing.py#L192 with https://github.com/pulp/pulp_deb/blob/main/pulp_deb/app/tasks/synchronizing.py#L513 .

Can you elaborate more on the strategy? Do we handle the namespace as if no avatar existed, or do we turn the artifact into on-demand?

File: pulp_ansible/app/tasks/collections.py [hunk 2]
```diff
@@ -1149,6 +1168,15 @@ def _pre_save(self, batch):
...
                 d_content.content.namespace = namespace
+                if d_content.d_artifacts:
+                    da = d_content.d_artifacts[0]
+                    # Check to see if avatar failed to download, update metadata if so
+                    if da.deferred_download:
+                        d_content.d_artifacts = None
+                        d_content.content.avatar_sha256 = None
+                    # Check to see if upstream didn't have avatar_sha256 set
+                    elif d_content.content.avatar_sha256 is None:
+                        d_content.content.avatar_sha256 = da.artifact.sha256
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collections.py:1179]: The name of this class does not match anymore.

File: pulp_ansible/app/tasks/collections.py [hunk 3]
```diff
@@ -1074,6 +1074,25 @@ async def run(self):
             pb.total = pb.done
 
 
+class DeclarativeNamespaceAvatar(DeclarativeArtifact):
```
Comment (mdellweg) [inline pulp_ansible/app/tasks/collections.py:1077]: Maybe give it a more functional name (like the DeclarativeFailsafeArtifact).

File: pulp_ansible/app/tasks/collections.py (hunk 1 above)
Comment (gerrod3) [reply pulp_ansible/app/tasks/collections.py:1090]: It is very similar. The current strategy I am using is to handle the namespace as if no avatar existed. In the custom content saver stage I modify the namespace's metadata before save if the avatar failed to download.

File: pulp_ansible/app/tasks/collections.py (hunk 2 above)
Comment (mdellweg) [reply pulp_ansible/app/tasks/collections.py:1179]: It's called `CollectionContentSaver`, yet it manipulates all types of content that come flying by (not your fault, but we could take the chance and make it a `AnsibleContentSaver`.).

File: pulp_ansible/app/tasks/collections.py (hunk 1 above)
Comment (gerrod3) [reply pulp_ansible/app/tasks/collections.py:1090]: Probably, we could design it to accept a list of exceptions during download that would be failsafe.

**Review by mdellweg — APPROVED** [review]

**Comment by mdellweg** [comment]
AAP-14090

**Pull request merged**